### PR TITLE
[all] Update issofa bugfix

### PR DESCRIPTION
**Pull request by hugtalbot**
Manual merge to close PR #218 
Only one commit was not integrated: (MechanicalMatrixVisitor and MechanicalOperations) fix if using a Linear Solver, projective constraints were wrongly applied when a force field is in a child node

-- Notes:

This PR will be further discussed during STC#3.

### Cleans
- BaseObject: Output message in serr for required datas. Error word should be reserved for messages that will make the application fail.
- RayTriangleVisitor and SceneLoaderFactory: clean warnings
- ParticlesRepulsionForceField: Add empty implementation of addKToMatrix to get rid of console warnings.

### Bugfixes
- AttachConstraint: fix crash when indices are resized
- BTDLinearSolver: fix constant log + operator new [] vs operator delete mismatch
- CatmullRomSplineMapping: fix applyJ compilation
- ConstantForceField: fix avoid crashes in draw when trying to apply a constantforcefield to a non existing point
- DistanceGrid: fix incorrect use of serr
- GeneralConstraintSolver: fix incorrect parent class in SOFA_CLASS macro
- IndicesFromValues: fix use getTemplateName to avoid problems of links with other data
- EulerImplicitSolver: fix Disable calls to solveConstraint from EulerImplicitSolver by default as it is not needed unless an very specific solver is added and it currently crashes in other cases + wrong argument order in calls to `AdvancedTimer::stepNext()`
- FixedConstraint and PartialFixedConstraint: fix "fixed FixedConstraint & PartialFixedConstraint for size-varying mechanical objects".
- MechanicalObject: fix crash in debug with null pointer
- Mass: fix remove error logging in Mass method that are inherited from Forcefield API and that are not relevant for Mass
- Mesh2PointMechanicalMapping: fix constraints ApplyJT
- Metis(gk_arch.h): fix compilation with Visual Studio 2015
- ParticlesRepulsionForceField and RepulsiveSpringForceField: fix avoid division by 0 on repulsion force fields
- PersistentContactBarycentricMapping: fix init variables in constructor
- RestShapeSpringsForceField: fix Runtime stiffness tunning was not working + optimisation and cleaning
- SkinningMapping: fix compilation of SofaRigid
- SofaViewer: fix crashes when current camera of pick-handler is NULL
- SurfacePressureForceField: fix volume computation formula
- TaitSurfacePressureForceField: fix contribution to the stiffness matrix for the second component df = P+dN in TaitSurfacePressureForceField
- TopologicalMapping: fix display error messages when a TopologicalMapping failed to be created
- TriangularFEMForceFieldOptim: fix principal values ordering with input matrix is diagonal + uninitialized value warning
- VisualModelImpl: fix wrong object (triangles) called when adding quads in handleTopologyChange()
- VoxelGridLoader: `std::unique` result was not used

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

File: SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl [hunk 1]
```diff
@@ -1524,30 +1535,42 @@ const Data<typename MechanicalObject<DataTypes>::VecCoord>* MechanicalObject<Dat
...
     }
 }
 
 template <class DataTypes>
 Data<typename MechanicalObject<DataTypes>::VecDeriv>* MechanicalObject<DataTypes>::write(core::VecDerivId v)
 {
+    if (v.isNull())
+    {
+        msg_error() << "Accessing null VecDeriv";
+    }
```
Comment (guparan) [inline SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl]: What should happen to v if `v.isNull()` ? This block does not pass in scene tests.

**Comment by epernod** [comment]
[ci-build][with-regression-tests] :+1:

File: SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl (hunk 1 above)
Comment (epernod) [reply SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl]: I tried to understand but...
So when v.isNull, this means that v.index == 0.

So in the next lines, it enters this block for the first time and create a vector.
```
if (vectorsDeriv[v.index] == NULL)
{
        vectorsDeriv[v.index] = new Data< VecDeriv >;
        vectorsDeriv[v.index]->setName(v.getName());
        vectorsDeriv[v.index]->setGroup("Vector");
```
For example In the scene: examples/Components/collision/RuleBasedContactManager.scn this error happened when constraints are created here:
```
template<class DataTypes>
void PairInteractionConstraint<DataTypes>::storeLambda(const ConstraintParams* cParams, MultiVecDerivId res, const sofa::defaulttype::BaseVector* lambda)
{
    if (cParams)
    {
        storeLambda(cParams, *res[mstate1.get(cParams)].write(), *res[mstate2.get(cParams)].write(), *cParams->readJ(mstate1), *cParams->readJ(mstate2), lambda);
    }
}
```

but the scene seems to act normally... (it moves)

Comment (epernod) [reply SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl]: I find it logical that a new Vecid is created when contact happened for the first time, but I don't understand why index is null.

**Review by epernod — CHANGES_REQUESTED** [review]

**Comment by epernod** [comment]
ok could you open an issue regarding the null vector to keep track of this bug. (maybe just a copy past of the lines)
And my comments regarding the investigation. You could even assign it to me! hurry up before I change my mind.

**Review by epernod — APPROVED** [review]

File: SofaKernel/framework/sofa/core/Mapping.inl [hunk 2]
```diff
@@ -127,7 +127,7 @@ sofa::defaulttype::BaseMatrix* Mapping<In,Out>::createMappedMatrix(const behavio
     sofa::defaulttype::BaseMatrix* result;
     if( !this->areMatricesMapped() )
     {
-        sout << "Mapping::createMappedMatrix() this mapping do not support matrices building. Set mapMatrices to true" << getClassName() << sendl;
+        msg_warning() << "Mapping::createMappedMatrix() this mapping do not support matrices building. Set mapMatrices to true" << getClassName();
```
Comment (damienmarchal) [inline SofaKernel/framework/sofa/core/Mapping.inl:130]: To be super pinchy (merge it witjout taking that into account)...
There is no need to add "Mapping::createMappedMatrix()" to the message as the emitting point is already in the msg_warning() macro. (additionally it make no sense to send to the user any details about c++ stuff)

File: modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl [hunk 3]
```diff
@@ -485,9 +590,37 @@ void Mesh2PointMechanicalMapping<TIn, TOut>::applyJT(const core::ConstraintParam
...
+                                fx = topoMap->getHexaBaryCoords()[i][0];
+                                fy = topoMap->getHexaBaryCoords()[i][1];
+                                fz = topoMap->getHexaBaryCoords()[i][2];
+                                err = false;
+                                break;
+                            }
+                        }
+                        if (err)
+                        {
+                            serr << "Error in " << __FUNCTION__ << " : wrong source hexa / destination point association" << sendl;
```
Comment (damienmarchal) [inline modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl:620]: miaouu

File: modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl [hunk 4]
```diff
@@ -470,9 +547,37 @@ void Mesh2PointMechanicalMapping<TIn, TOut>::applyJT(const core::ConstraintParam
...
+                                fx = topoMap->getTetraBaryCoords()[i][0];
+                                fy = topoMap->getTetraBaryCoords()[i][1];
+                                fz = topoMap->getTetraBaryCoords()[i][2];
+                                err = false;
+                                break;
+                            }
+                        }
+                        if (err)
+                        {
+                            serr << "Error in " << __FUNCTION__ << " : wrong source tetra / destination point association" << sendl;
```
Comment (damienmarchal) [inline modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl:578]: mheuu

File: modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl [hunk 5]
```diff
@@ -456,8 +507,34 @@ void Mesh2PointMechanicalMapping<TIn, TOut>::applyJT(const core::ConstraintParam
...
+                            {
+                                fx = topoMap->getQuadBaryCoords()[i][0];
+                                fy = topoMap->getQuadBaryCoords()[i][1];
+                                err = false;
+                                break;
+                            }
+                        }
+                        if (err)
+                        {
+                            serr << "Error in " << __FUNCTION__ << " : wrong source quad / destination point association" << sendl;
```
Comment (damienmarchal) [inline modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl:535]: coincoin

File: modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl [hunk 6]
```diff
@@ -443,8 +468,34 @@ void Mesh2PointMechanicalMapping<TIn, TOut>::applyJT(const core::ConstraintParam
...
+                            {
+                                fx = topoMap->getTriangleBaryCoords()[i][0];
+                                fy = topoMap->getTriangleBaryCoords()[i][1];
+                                err = false;
+                                break;
+                            }
+                        }
+                        if (err)
+                        {
+                            serr << "Error in " << __FUNCTION__ << " : wrong source triangle / destination point association" << sendl;
```
Comment (damienmarchal) [inline modules/SofaTopologyMapping/Mesh2PointMechanicalMapping.inl:496]: msg_warning..

**Comment by damienmarchal** [comment]
Ok, zouh.

**Pull request merged**